index.d.ts

```diff
@@ -0,0 +1,42 @@
+declare namespace closeWithGrace {
+  interface Options {
+    /**
+     * The numbers of milliseconds before abruptly close the process
+     * @default 10000
+     */
+    delay: number
+  }
+
+  type Signals = 'SIGTERM' | 'SIGINT';
+
+  type CloseWithGraceCallback = (
+    { err, signal, manual }: { err: Error, signal: Signals, manual: boolean },
+    cb: (error?: Error) => void
```

Comment:
the callback is optional. In that case the function should return a promise. If there is a callback, the function must not return a promise.

Reply:
You are totally right but can't find a way to achieve that 😞. I tried overloading but seems like creating a union of the  types 

```ts
type asyncAddCb = (result: number) => void
// define all valid function signatures
function asyncAdd (a: number, b: number): Promise<number>
function asyncAdd (a: number, b: number, cb: asyncAddCb): void

function asyncAdd (a: number, b: number, cb?: asyncAddCb) {
  const result = a + b

  return Promise.resolve(result);
}
```

also gave a try to this syntax 
`typeof cb extends undefined ? Promise<void> : void;` but doesn't seem to be correct 😕

Not sure if someone is a ts expert to share some knowledge here 😅

Reply:
At least let's not test the bad behavior.

Reply:
@gkampitakis actually you overload works fine as you can see here on the [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAhgziAdgYwIIBN0GEBGUC8UAFAE4RwCuANsAFxSIUC2OEJAlAQHxQBuA9gEt0AKABmFFMEH9EsBCgzpiMeoxZsANFBxrmrDvQAKJfk0FwIAHnUGu4ycmmz5SNJhV6NJbbob6tKGQ-eDclXHZ6AWERBykZOVDFDyJVf29fLwNtYIB+eiT3bBxOAG8RKCDZOGAoMkoaAlgoAGodCqhBMWJgso7KyuDScmpgdn6BsmAKEkQOgF8OqZm5EzMLCAA6ev4qXghhhrGAbhFF1IVkKAl4lyI+garEGrqRxsIYAHcYQVrCpSIAFZNAA2caPZDVXZbKj8ADmh1G41iKIh1Vq9VGACYmv9MEDQdpETROPgeOVHmjntDNrCEZiSRN5uNmfcREA).

<img width="399" alt="image" src="https://user-images.githubusercontent.com/1620916/109469518-69813a00-7a6e-11eb-8043-f74611d275ea.png">

<img width="649" alt="image" src="https://user-images.githubusercontent.com/1620916/109469574-7ef66400-7a6e-11eb-9d77-53ed7f838891.png">

Reply:
When I try to do this though 
```
type CloseWithGraceCallback = {
    (
      { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
      cb: (error?: Error) => void
    ): void
    (
      { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
    ): Promise<void>
    (
      { err, signal, manual }: { err?: Error, signal?: Signals, manual?: boolean },
      cb?: (error?: Error) => void
    ): Promise<void> | void
  }
```

I get those two errors from the callback case 
<img width="1346" alt="Screenshot 2021-03-01 at 09 37 36" src="https://user-images.githubusercontent.com/50364739/109479509-5b351d00-7a72-11eb-8601-365f762f010b.png">
<img width="1345" alt="Screenshot 2021-03-01 at 09 37 53" src="https://user-images.githubusercontent.com/50364739/109479515-5ec8a400-7a72-11eb-8260-91e18be23828.png">